staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	"k8s.io/apimachinery/pkg/util/httpstream/wsstream"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	constants "k8s.io/apimachinery/pkg/util/portforward"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/klog/v2"
+)
+
+// TunnelingHandler is a handler which tunnels SPDY through WebSockets.
+type TunnelingHandler struct {
+	// UpgradeAwareHandler used to communicate to upstream SPDY.
+	upgradeHandler http.Handler
+	// Server-side of upgraded websocket connection
+	tunnelingConnection net.Conn
+	// MaxBytesPerSec throttles stream Reader/Writer if necessary
+	MaxBytesPerSec int64
+}
+
+func NewTunnelingHandler(upgradeHandler http.Handler, maxBytesPerSec int64) *TunnelingHandler {
+	return &TunnelingHandler{
+		upgradeHandler: upgradeHandler,
+		MaxBytesPerSec: maxBytesPerSec,
+	}
+}
+
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.Infoln("TunnelingHandler ServeHTTP")
+	clone := utilnet.CloneRequest(req)
+
+	// First, terminate the websocket connection, creating the net.Conn
+	// that will be used to tunnel SPDY.
+	klog.Infoln("Upgrading, terminating websocket connection...")
+	var upgrader = gwebsocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
+			return true // Accepting all requests
+		},
+		Subprotocols: []string{
+			constants.PortForwardV2Name,
```

Comment:
can define this as the server-side version of https://github.com/kubernetes/kubernetes/pull/123413/files#r1497897438, e.g. `"SPDY/3.1+"+constants.PortForwardV1Name`

Reply:
I don't think we need to do anything that combines the SPDY/WebSockets headers. This should be just the websocket v2 portforward protocol as sent by the tunneling dialer. After the websocket connection is created in the client (using standard websocket headers, including v2) it is wrapped in the `TunnelingConnection` which implements `net.Conn`. This is what is passed into the creation of the SPDY client, transparently tunneling all SPDY connection calls.